### PR TITLE
Add teacher forcing demo test for Deepseek 671B model

### PR DESCRIPTION
**Pull request by yalrawwashTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/33968

### Problem description
No accuracy check for Deepseek 671B parameter model

### What's changed
Add teacher forcing demo test for Deepseek 671B model to ensure deterministic output

### Checklist
[(Galaxy) DeepSeek tests](https://github.com/tenstorrent/tt-metal/actions/runs/20780128967)

### CODE REVIEW

File: models/demos/deepseek_v3/demo/test_demo_teacher_forced.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    # gt_tokens = Phase 1's decode outputs
+    # With the generator fix, teacher forcing starts at decode step 0:
+    # - Decode 0: input is natural prefill output, predict and collect, force gt[0]
+    # - Decode 1: input is gt[0], predict and collect, force gt[1]
+    # - ...
+    full_sequence = raw_prompt_tokens + generated_tokens
+    reference_tokens_tensor = torch.tensor(full_sequence, dtype=torch.long).unsqueeze(0)
+
+    temp_reference_file = Path("/tmp/deepseek_v3_teacher_forcing_test.refpt")
```
[the remark this answers is not in the document]
Comment (yieldthought) [reply models/demos/deepseek_v3/demo/test_demo_teacher_forced.py]: This .refpt file should be small enough to put in source control, no? That saves us running the reference model at all and means we can also run the full model.

File: models/demos/deepseek_v3/tt/generator.py [hunk 2]
```diff
@@ -530,11 +538,11 @@ def generate(
             logger.info(f"First sampled token: {self.tokenizer.decode(token_value, skip_special_tokens=True)}")
 
             positions = torch.zeros(self.batch_size, dtype=torch.int32) + lengths
-            # If teacher forcing is enabled, collect the model's predicted token and force GT for next step (single prompt)
-            if teacher_forcing is not None:
-                # Only enforce for the first user to keep scope minimal
-                forced = teacher_forcing.collect_predicted_tokens(int(next_tokens[0].item()))
-                next_tokens[0] = int(forced)
+            # NOTE: Do NOT apply teacher forcing here after prefill.
+            # The model's natural first token (A') should be used as input to decode step 0.
```
Comment (yieldthought) [inline models/demos/deepseek_v3/tt/generator.py]: This makes no sense to me - why would we want to use the model's first token for step 0? If it's different to the teacher's then the whole sequence is thrown off.

Comment (yieldthought) [reply models/demos/deepseek_v3/tt/generator.py]: To be clear, we should include the student and teacher comparison to test whether A' and A are the same, but the student should always run decode(A) and not decode(A')

File: models/demos/deepseek_v3/demo/generate_teacher_forced_file.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    store them in a file that will be used later by the pytest.
+
+    The saved file includes:
+      - reference_tokens: full prompt+generated sequence (1 x T tensor)
+      - tf_prompt_len: length of the prompt in tokens
+      - max_new_tokens: used when the reference was generated
+      - prompt: the text prompt used for generation
+    """
+
+    print("\n=== Phase 1: Generate reference tokens with TT model ===")
```
Comment (yieldthought) [inline models/demos/deepseek_v3/demo/generate_teacher_forced_file.py]: The point is to run the huggingface reference model to generate our reference tokens - and to do so ahead of time and save the small .refpt so we don't rerun it every time we run a CI test.

I'm not sure what comparing our model against itself is supposed to test here?

**Review by yieldthought — CHANGES_REQUESTED** [review]

File: models/demos/deepseek_v3/demo/generate_teacher_forced_file.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        loaded_weight = loaded_weight.to(dtype=target_dtype, device=target_device)
+        if tensor.dtype != target_dtype:
+            tensor.data = tensor.data.to(target_dtype)
+        tensor.data = loaded_weight
+        del loaded_weight
+        return tensor
+
+    def add_dynamic_weight_loading_hooks_with_dtype(
+        module, weights_dict, lazy_modules=["DeepseekV3Attention", "DeepseekV3MLP"], model_name=""
+    ):
```
Comment (yalrawwashTT) [inline models/demos/deepseek_v3/demo/generate_teacher_forced_file.py:107]: this file is how I generated the hugging face reference, but is not needed for the final PR. just need to test on 61 layers, but blocked by cache issue atm

**Comment by yalrawwashTT** [comment]
```
[1,0]<stdout>: === Comparison ===
[1,0]<stdout>: Prompt Text:
[1,0]<stdout>: <｜begin▁of▁sentence｜><｜User｜>What is the capital of France? Please provide a brief answer.<｜Assistant｜>
[1,0]<stdout>: 
[1,0]<stdout>: HF Reference Output (Tokens: 32):
[1,0]<stdout>: <think>
[1,0]<stdout>: Okay, the user is asking for the capital of France and wants a brief answer. 
[1,0]<stdout>: 
[1,0]<stdout>: Hmm, this is a very basic geography question - almost
[1,0]<stdout>: 
[1,0]<stdout>: HF Reference Token IDs:
[1,0]<stdout>: [128798, 201, 33001, 14, 270, 3967, 344, 13070, 362, 270, 6102, 294, 8760, 305, 10059, 260, 9708, 3287, 16, 1004, 42, 6002, 14, 566, 344, 260, 1855, 6569, 37153, 3417, 565, 4975]
[1,0]<stdout>: 
[1,0]<stdout>: TT Forced Output (Tokens: 32):
[1,0]<stdout>: <think>
[1,0]<stdout>: Okay, the user is asking for the capital of France and wants a brief answer. 
[1,0]<stdout>: 
[1,0]<stdout>: Hmm, this is a very basic geography question - almost
[1,0]<stdout>: 
[1,0]<stdout>: TT Forced Token IDs:
[1,0]<stdout>: [128798, 201, 33001, 14, 270, 3967, 344, 13070, 362, 270, 6102, 294, 8760, 305, 10059, 260, 9708, 3287, 16, 1004, 42, 6002, 14, 566, 344, 260, 1855, 6569, 37153, 3417, 565, 4975]
[1,0]<stdout>: 
[1,0]<stdout>: TT Predicted Output (Tokens: 32):
[1,0]<stdout>: <think><think>H, the The asked asking " " "​​​​​�​�​�​​​​​​​​​​​
[1,0]<stdout>: 
[1,0]<stdout>: TT Predicted Token IDs:
[1,0]<stdout>: [128798, 128798, 42, 14, 270, 455, 4869, 13070, 582, 582, 582, 35020, 35020, 35020, 35020, 35020, 300, 35020, 300, 35020, 300, 35020, 35020, 35020, 35020, 35020, 35020, 35020, 35020, 35020, 35020, 35020]
[1,0]<stdout>: 
[1,0]<stdout>: HF Model Top-5 Predictions (for generated token positions):
[1,0]<stdout>:   Position 16: top5=['<think>', '</think>', '<｜place▁holder▁no▁797｜>', '<｜place▁holder▁no▁793｜>', '<｜place▁holder▁no▁794｜>'], actual='<think>' (id=128798) ✓
[1,0]<stdout>:   Position 17: top5=['\n', '\n\n', ' \n', 'H', '\n\n\n'], actual='\n' (id=201) ✓
[1,0]<stdout>:   Position 18: top5=['Okay', 'H', 'The', 'Alright', 'We'], actual='Okay' (id=33001) ✓
[1,0]<stdout>:   Position 19: top5=[',', ' the', ' let', ' user', ' so'], actual=',' (id=14) ✓
[1,0]<stdout>:   Position 20: top5=[' the', ' user', ' let', ' I', ' so'], actual=' the' (id=270) ✓
[1,0]<stdout>: ==================
[1,0]<stdout>: 
[1,0]<stdout>: 
[1,0]<stdout>: Top-1 accuracy: 0.1250 (12.50%)
[1,0]<stdout>: Top-5 accuracy: 0.2812 (28.12%)
```

File: models/demos/deepseek_v3/demo/generate_teacher_forced_file.py
Comment (yalrawwashTT) [inline models/demos/deepseek_v3/demo/generate_teacher_forced_file.py:1]: This is what I used to generate the .refpt file, but once that is generated, this is no longer needed.

**Review by yieldthought — APPROVED** [review]
